pii_recognition/registers/registry.py

```diff
@@ -10,8 +11,18 @@ def __init__(self):
 
     @abstractmethod
     def add_predefines(self):
+        """Use add_item method to put objects to registry."""
         ...
 
-    def add_item(self, item: T):
-        name = getattr(item, "__name__")
-        self[name] = item
+    def add_item(self, item: T, name: Optional[str] = None):
```

Comment:
I think what you want is 
```
class Registry(dict, Generic[T], metaclass=ABCMeta):
    ...

    def add_item(self, item: Type[T], name: Optional[str] = None):
        ...
```
and then the `TokeniserRegistry` can inherit from `Registry[Tokeniser]` where `Tokeniser` is an abstract parent class for all tokenisers.
and

Reply:
Fixed.

Reply:
Also, if we use generic class, would it still make sense to have it as abstract class inheriting ABCMeta? Usage of generic will allow this
```
tokeniser_registry = registry[Tokeniser]()  # registry[Tokeniser] instance already
tokeniser_registry.add_item(Tokeniser_1)
tokeniser_registry.add_item(Tokeniser_2)
```
we don't need to actually create a child class that inherits from `Registry` , i.e., `TokeniserRegistry(Registry)`.

Reply:
ah.. yes, doesn't need to be abstract class

Reply:
but the `tokeniser_registry.add_item(Tokeniser_1)` line should be in the module which defines `Tokeniser_1` rather than the module with `tokeniser_registry`

Reply:
I didn't think it through at the first time. If we do so as mentioned, how can we use registry when needed if there's no a centralised manager for instantiation? An illustration of example doing in place registration using `add_item`, which I think is what you mean
```
# in registers/__init__.py file
from .registry import Registry
from tokenisation.tokeniser import Tokeniser

tokeniser_registry = Registry[Tokeniser]()
``` 

```
# tokenisers/tokeniser_1.py file
from registers import tokeniser_registry

class Tokeniser_1(Tokeniser): ...
tokeniser_registry.add(Tokeniser_1)
```
but how to use `tokeniser_registry`, for example, within a pipeline like the below?
```
from register import tokeniser_registry

tokeniser_1 = tokeniser_registry["Tokeniser_1"] # this does not work no specific tokeniser registered
```
What you think of [this](https://github.com/tensorflow/hub/blob/2bd67f1df78c7d465254c3e21e96d8782b37d653/tensorflow_hub/config.py), in this example they use a config file to manager and register all predefines. But I feel it would be simpler to do registration wherever `tokeniser_registry` initialised.